libsel4/tools/generator.py

```diff
@@ -0,0 +1,612 @@
+
+                if param_dir == "in" or param_type.pass_by_reference():
+                    param_description = param.getAttribute("description")
+                    if not param_description:
+                        param_description_element = param.getElementsByTagName("description")
+                        if param_description_element:
+                            param_description_text = get_xml_element_content_with_xmlonly(
+                                param_description_element[0])
+                            param_description = normalise_text(param_description_text)
+
+                    comment_lines.append("@param[%s] %s %s " %
+                                         (param_dir, param_name, param_description))
+
+            method_return_description = method.getElementsByTagName("return")
+            if method_return_description:
+                comment_lines.append("@return @xmlonly %s @endxmlonly" %
+                                     get_xml_element_contents(method_return_description[0]))
+            else:
+                # no return documentation given - default to something sane
+                if is_result_struct_required(output_params):
+                    comment_lines.append("@return @xmlonly @endxmlonly")
+                else:
+                    comment_lines.append("@return @xmlonly <errorenumdesc/> @endxmlonly")
+
+            for error in method.getElementsByTagName("error"):
+                error_name = error.getAttribute("name")
+                error_description = error.getAttribute("description")
+                if not error_description:
+                    error_description_element = error.getElementsByTagName("description")
+                    if error_description_element:
+                        error_description_text = get_xml_element_content_with_xmlonly(
+                            error_description_element[0])
+                        error_description = normalise_text(error_description_text)
+                comment_lines.append("@retval %s %s " % (error_name, error_description))
+
+            # split each line on newlines
+            comment_lines = reduce(operator.add, [l.split("\n") for l in comment_lines], [])
+
+            methods.append((interface_name, method_name, method_id, input_params,
+                            output_params, method_condition, comment_lines))
+
+    return (methods, structs, api)
+
+
+class Generator:
+    def __init__(self):
+        self.contents = []
+
+    def init_data_types(self):
+        raise NotImplementedError()
+
+    def init_arch_types(self):
+        raise NotImplementedError()
+
+    def construction(self, expr, param):
+        raise NotImplementedError()
+
+    @staticmethod
+    def generate_param_list(input_params, output_params):
+        raise NotImplementedError()
+
+    def configure(self, arch, wordsize, xml_files, buffer, mcs):
+        # Ensure architecture looks sane.
+        if arch not in WORD_SIZE_BITS_ARCH.keys():
```

Comment:
This would look up the `arch` object then. Choices for `wordsize` and `mcs` might be something that the architecture has to support then or this fails.